### ISSUE TITLE
CUID vs BSON ObjectID ?

### ISSUE THREAD

**Issue by kevbaker** [ericelliott/cuid#162]
I’m evaluating CUID vs BSON ObjectID. ObjectID is what MongoDB is using for unique ID’s. Are there specific reasons where a CUID would make more sense? Uniqueness maybe?

BTW: I’m favoring both over UUID do to the somewhat sequential generation to help with indexing performance.

**Comment by gajus** [comment]
@kevbaker Did you find an answer?

I am curious to know as well.

**Comment by ericelliott** [comment]
ObjectID is very similar, but has no notion of host fingerprint, which could cause collisions when multiple ids are generated in the same millisecond on different machines.

The problem is that [not all random number generators provide good collision resistance](https://hackaday.com/2015/12/28/v8-javascript-fixes-horrible-random-number-generator/). I've seen cross-host id collisions in PHP, Ruby, Java, and various database ID generators (I have not battle tested Mongo's implementation with tens of thousands of simultaneous users, sorry), etc. Many old SQL database engines defaulted to an auto-incrementing integer IDs, which guarantees cross-host collisions!

TL;DR - You may be more likely to see a collision using ObjectID, but it's going to be very implementation-dependent and I have not battle-tested ObjectID. That said, it has been quite a while since I heard of a collision in a full cuid. It's used as the default primary key in some database engines, and in every production system I've needed client side IDs for since I created it. Apps with tens of millions of users creating billions of id-keyed entities.
